### ISSUE TITLE
Applying heatmap overlay in 3D

### ISSUE THREAD

**Issue by bsolano** [MECLabTUDA/M3d-Cam#5]
Hi.  Documentation does not say how to apply a 3D heatmap.  For instance, I am inputing an MRI (87x75x100) to my model and the gcampp injected returns a heatmap that is (2x3x5).  It is not clear how to apply that.  Also, if I use ggcamp I obtain an axial result but my input was sagittal.  It is not a big issue, it's just a matter of rotating, but it was not intuitive. Therefore I am not sure if the heatmap returned by gcampp is also rotated.  I have no clue on how to use it on 3D data.  Best regards.

**Comment by Karol-G** [comment]
Hi,

>Documentation does not say how to apply a 3D heatmap. For instance, I am inputing an MRI (87x75x100) to my model and the gcampp injected returns a heatmap that is (2x3x5). It is not clear how to apply that.

Most models use some kind of bottleneck architecture. So the layers in the middle are smaller than the layer at the input/output. If you extract attention maps at the middle of these models the attention maps will be smaller as well.  Simply rescale your attention maps if you want to superimpose the 3D attention maps on your input. For 2D this is done automatically, but I did not implement that yet for 3D, so you need to do it yourself.

> Also, if I use ggcamp I obtain an axial result but my input was sagittal. It is not a big issue, it's just a matter of rotating, but it was not intuitive.

Hmm, I will have a look into this within the next days.

Best
Karol

**Comment by dxs66** [comment]
> Hi. Documentation does not say how to apply a 3D heatmap. For instance, I am inputing an MRI (87x75x100) to my model and the gcampp injected returns a heatmap that is (2x3x5). It is not clear how to apply that. Also, if I use ggcamp I obtain an axial result but my input was sagittal. It is not a big issue, it's just a matter of rotating, but it was not intuitive. Therefore I am not sure if the heatmap returned by gcampp is also rotated. I have no clue on how to use it on 3D data. Best regards.

Hi, I am using the M3d-Cam to apply a 3D heatmap.But now i met some problems that i can get the output dirs,but there are no  3D heatmap in it. My usage is as follows.
![image](https://user-images.githubusercontent.com/29002324/116056135-039ce180-a6b0-11eb-8440-0c7c986f67ce.png)
![微信图片_20210426165626](https://user-images.githubusercontent.com/29002324/116056467-67bfa580-a6b0-11eb-868c-a7f329ff0fcb.png)
can you send your codes to me to solve this problem? Thank you !

**Comment by Karol-G** [comment]
Hi,

are you able to generate attention maps with the default inject arguments like this?

`model = medcam.inject(model, output_dir="attention_maps", save_maps=True)`

If yes, add more of your current arguments successively until the bug occurs.
Can you also send me the code were you call the forward of your model?

Best
Karol

**Comment by dxs66** [comment]
> Hi,
> 
> are you able to generate attention maps with the default inject arguments like this?
> 
> `model = medcam.inject(model, output_dir="attention_maps", save_maps=True)`
> 
> If yes, add more of your current arguments successively until the bug occurs.
> Can you also send me the code were you call the forward of your model?
> 
> Best
> Karol

Hello,
 I have tried your methods , but there are still no results. The outcome is as follows

**Comment by dxs66** [comment]
> Hi,
> 
> are you able to generate attention maps with the default inject arguments like this?
> 
> `model = medcam.inject(model, output_dir="attention_maps", save_maps=True)`
> 
> If yes, add more of your current arguments successively until the bug occurs.
> Can you also send me the code were you call the forward of your model?
> 
> Best
> Karol

Hello,
I have tried your method, but there are still no results.  The outcome are as follows
![image](https://user-images.githubusercontent.com/29002324/116075146-2980b100-a6c5-11eb-9223-923736247931.png)
![image](https://user-images.githubusercontent.com/29002324/116075239-487f4300-a6c5-11eb-85ea-2237b187acbe.png)
The forward of my model is as follows
![image](https://user-images.githubusercontent.com/29002324/116075336-6a78c580-a6c5-11eb-96e3-e63b21f5f67e.png)

**Comment by dxs66** [comment]
![image](https://user-images.githubusercontent.com/29002324/116076164-7749e900-a6c6-11eb-8ec6-7a0e4575d26a.png)

**Comment by Karol-G** [comment]
Ah, I get the problem now.
Medcam backups & replaces the models original forward with its own which should be called just "forward".
This wont work in your case as your forward is called "forward_global" and "forward_local", so medcam can't find them.
Try to rename the forward that you are using to "forward".

**Comment by bsolano** [comment]
Hello!

> > Documentation does not say how to apply a 3D heatmap. For instance, I am inputing an MRI (87x75x100) to my model and the gcampp injected returns a heatmap that is (2x3x5). It is not clear how to apply that.
> 
> Most models use some kind of bottleneck architecture. So the layers in the middle are smaller than the layer at the input/output. If you extract attention maps at the middle of these models the attention maps will be smaller as well. Simply rescale your attention maps if you want to superimpose the 3D attention maps on your input. For 2D this is done automatically, but I did not implement that yet for 3D, so you need to do it yourself.

I am sorry for this very late response.  I was using this code:

`class_names = ['CN','EMCI','MCI','LMCI','AD']`  
`model = densenet121(channels=1, num_classes=len(class_names), drop_rate=0.7).to(device)`  
`model = medcam.inject(model, output_dir="attention_maps", backend='ggcam', label=4, save_maps=True, metric='wioa')`  
`model = torch.nn.DataParallel(model).to(device)`  
`model.load_state_dict(torch.load('../../Alzheimer-ResNets/results-6-2-2020-extended/cuda-epoch-109-alzheimer-densenet121.pth'))`

According to the documentation layer default value is 'auto'.  And the library was saying "Selected module layer: features".  Therefore, I believe I was not using a middle layer.  Am I right?

> > Also, if I use ggcamp I obtain an axial result but my input was sagittal. It is not a big issue, it's just a matter of rotating, but it was not intuitive.
> 
> Hmm, I will have a look into this within the next days.

Did you look into this?

I appreciate very much your work.  Thank you.

Best regards,

Braulio  J. Solano-Rojas

**Comment by nasir3843** [comment]
Hi bsolano 
Did you manage to solve the problem of overlaying attention maps on the 3D input volume. I am facing the same problem. In my case I get 6x6x6 sized attention maps from the last convolution layer. My input volume to the network is 110x110x110. I am not sure how to overlay these attention maps on the input volume of images. Your help in this regard will be appreciated.
Thank you
